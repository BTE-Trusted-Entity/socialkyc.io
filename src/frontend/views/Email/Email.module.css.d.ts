export const __esModule: true;
export const heading: string;
export const formInput: string;
export const inputError: string;
export const profile: string;
