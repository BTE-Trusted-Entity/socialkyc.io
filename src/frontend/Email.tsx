import { useCallback, useState } from 'react';
import { Route, Switch, useLocation } from 'react-router';
import { Link } from 'react-router-dom';
import cx from 'classnames';
import ky from 'ky';
import { StatusCodes } from 'http-status-codes';
import { IEncryptedMessage } from '@kiltprotocol/types';

import { getSession } from './utilities/session';
import { handleBeforeUnload } from './utilities/handleBeforeUnload';

export function Email(): JSX.Element {
  const { pathname } = useLocation();

  const [nameInput, setNameInput] = useState('');
  const [emailInput, setEmailInput] = useState('');

  const handleNameInput = useCallback((event) => {
    setNameInput(event.target.value);
  }, []);

  const handleEmailInput = useCallback((event) => {
    setEmailInput(event.target.value);
  }, []);

  const [email, setEmail] = useState('');

  const handleSubmit = useCallback(
    async (event) => {
      event.preventDefault();
      window.addEventListener('beforeunload', handleBeforeUnload);
      try {
        const session = await getSession();
        await session.listen(async (message) => {
          const result = await ky.post('/request-attestation', {
            json: message,
          });

          window.removeEventListener('beforeunload', handleBeforeUnload);

          if (result.status === StatusCodes.ACCEPTED) {
            console.log('Terms rejected');
          }

          if (result.status !== StatusCodes.OK) {
            console.log('Not attested');
            return;
          }

          setEmail(await result.text());
        });

        const json = {
          name: nameInput,
          email: emailInput,
          did: session.identity,
        };

        const message = (await ky
          .post('/quote', { json })
          .json()) as IEncryptedMessage;

        console.log('Message: ', message);

        await session.send(message);
      } catch (error) {
        console.error(error);
        window.removeEventListener('beforeunload', handleBeforeUnload);
      }
    },
    [nameInput, emailInput],
  );

  return (
    <li
      className={cx('expandableItem', {
        expanded: pathname === '/email',
      })}
    >
      <p className="itemLabel">
        <Switch>
          <Route path="/email">
            <Link to="/" type="button" className="button accordion opened" />
          </Route>
          <Route>
            <Link
              to="/email"
              type="button"
              className="button accordion closed"
            />
          </Route>
        </Switch>
        Email
      </p>
      <Route path="/email">
        {email ? (
          <div className="success">
            <p>
              Weâ€™ve sent an email to <strong>{email}</strong>
            </p>
            <p>Please check your inbox!</p>
            <Link to="/" type="button" className="button buttonPrimary">
              OK
            </Link>
          </div>
        ) : (
          <form className="emailForm" onSubmit={handleSubmit}>
            <label className="formLabel">
              Your full name
              <input
                className="formInput"
                onInput={handleNameInput}
                type="text"
                name="name"
                required
              />
            </label>

            <label className="formLabel">
              Your email address
              <input
                className="formInput"
                onInput={handleEmailInput}
                type="email"
                name="email"
                required
              />
            </label>

            <button
              type="submit"
              className="button buttonPrimary chooseIdentity"
              disabled={!nameInput || !emailInput}
            >
              Choose Sporran Identity
            </button>
          </form>
        )}
      </Route>
    </li>
  );
}
