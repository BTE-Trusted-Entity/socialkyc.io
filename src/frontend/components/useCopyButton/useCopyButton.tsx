import { MouseEventHandler, RefObject, useCallback, useState } from 'react';

import * as styles from './useCopyButton.module.css';

interface UseCopyButton {
  supported: boolean;
  handleCopyClick: MouseEventHandler;
  className: string;
  title: string;
  justCopied: boolean;
}

export function useCopyButton(
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>,
): UseCopyButton {
  const writeTextSupported =
    navigator.clipboard && 'writeText' in navigator.clipboard;
  const supported =
    writeTextSupported || document.queryCommandSupported('copy');

  const [justCopied, setJustCopied] = useState(false);
  const className = justCopied ? styles.copied : styles.copy;

  const handleCopyClick = useCallback(async () => {
    const input = inputRef.current;
    if (!input || !supported) {
      return;
    }

    if (writeTextSupported) {
      await navigator.clipboard.writeText(input.value);
    } else {
      input.select();
      document.execCommand('copy');
    }

    setJustCopied(true);
    setTimeout(() => {
      setJustCopied(false);
    }, 500);
  }, [inputRef, supported, writeTextSupported]);

  return {
    supported,
    handleCopyClick,
    className,
    title: 'Copy',
    justCopied,
  };
}
