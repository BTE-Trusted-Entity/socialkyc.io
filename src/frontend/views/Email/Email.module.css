.heading {
  composes: heading from '../../components/CredentialFlow/CredentialFlow.module.css';
  background-image: url('../Attester/email.svg');
}

.formInput {
  composes: input from '../../components/Input/Input.module.css';
  margin-bottom: 0;
}

.inputError {
  composes: error from '../../components/Input/Input.module.css';
}

.profile {
  display: grid;
  grid-template-areas: 'key value' 'key value';
  width: calc(250rem / 16);
  gap: 0.5rem;
  margin-bottom: 2rem;
}
