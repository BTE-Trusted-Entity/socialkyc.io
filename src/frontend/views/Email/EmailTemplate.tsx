import {
  FormEventHandler,
  Fragment,
  MouseEventHandler,
  useCallback,
  useState,
} from 'react';
import { Prompt } from 'react-router-dom';
import cx from 'classnames';

import * as flowStyles from '../../components/CredentialFlow/CredentialFlow.module.css';
import * as styles from './Email.module.css';

import { usePreventNavigation } from '../../utilities/usePreventNavigation';
import { Spinner } from '../../components/Spinner/Spinner';
import { Explainer } from '../../components/Explainer/Explainer';
import { expiryDate } from '../../utilities/expiryDate';
import { SigningErrorClosed } from '../../components/SigningErrorClosed/SigningErrorClosed';
import { DetailedMessage } from '../../components/DetailedMessage/DetailedMessage';
import { AttestationProcessAnchoring } from '../../components/AttestationProcessAnchoring/AttestationProcessAnchoring';
import { AttestationProcessReady } from '../../components/AttestationProcessReady/AttestationProcessReady';
import { AttestationErrorUnknown } from '../../components/AttestationErrorUnknown/AttestationErrorUnknown';
import { LinkBack } from '../../components/LinkBack/LinkBack';

export type AttestationStatus =
  | 'none'
  | 'requested'
  | 'confirming'
  | 'attesting'
  | 'ready'
  | 'error';

export type FlowError = 'closed' | 'expired' | 'unknown';

interface Props2 {
  status: AttestationStatus;
  processing: boolean;
  handleSubmit: FormEventHandler;
  handleBackup: MouseEventHandler;
  handleTryAgainClick: MouseEventHandler;
  setInputError: (error?: string) => void;
  inputError?: string;
  flowError?: FlowError;
}

export function EmailTemplate({
  status,
  processing,
  handleSubmit,
  handleBackup,
  handleTryAgainClick,
  setInputError,
  inputError,
  flowError,
}: Props2): JSX.Element {
  const [emailInput, setEmailInput] = useState('');

  const handleInput = useCallback(
    (event) => {
      setEmailInput(event.target.value);
      setInputError(undefined);
    },
    [setInputError],
  );

  const handleBlur = useCallback(() => {
    const trimmed = emailInput.trim();
    if (trimmed !== emailInput) {
      setEmailInput(trimmed);
    }
  }, [emailInput]);

  const handleKeyPress = useCallback((event) => {
    if (event.key === 'Enter') {
      event.target.blur();
    }
  }, []);

  const preventNavigation = usePreventNavigation(
    processing || ['confirming', 'attesting'].includes(status),
  );

  return (
    <section
      className={cx(flowStyles.container, {
        [flowStyles.processing]: processing,
      })}
    >
      {processing && <Spinner />}

      <h1 className={styles.heading}>Email Address Attestation</h1>

      <Prompt
        when={preventNavigation}
        message="The email attestation process has already started. Are you sure you want to leave?"
      />

      <Explainer>
        After entering your email address, please choose an Identity in your
        wallet to associate with your email credential. We will email you a link
        so that we can attest your credential.
      </Explainer>

      {status === 'none' && (
        <form onSubmit={handleSubmit}>
          <label>
            Your email address
            <input
              className={styles.formInput}
              onInput={handleInput}
              value={emailInput}
              type="email"
              name="email"
              required
              maxLength={100}
              onBlur={handleBlur}
              onKeyPress={handleKeyPress}
            />
          </label>

          <output className={styles.inputError} hidden={!inputError}>
            {inputError}
          </output>

          <p>Validity: one year ({expiryDate})</p>

          {flowError === 'closed' && <SigningErrorClosed />}

          <button
            type="submit"
            className={flowStyles.ctaButton}
            disabled={!emailInput}
          >
            Continue in wallet
          </button>
        </form>
      )}

      {status === 'requested' && (
        <DetailedMessage
          icon="spinner"
          heading="Attestation process:"
          message="Email sent"
          details={`Email sent to ${emailInput}. Please check your inbox and spam folder and click the link.`}
        />
      )}

      {status === 'attesting' && <AttestationProcessAnchoring />}

      {status === 'ready' && (
        <Fragment>
          <AttestationProcessReady />

          <button
            className={flowStyles.ctaButton}
            type="button"
            onClick={handleBackup}
          >
            Show credential in wallet
          </button>
        </Fragment>
      )}

      {flowError === 'unknown' && <AttestationErrorUnknown />}

      {flowError === 'expired' && (
        <DetailedMessage
          icon="exclamation"
          heading="Attestation error:"
          message="This link has expired."
        />
      )}

      {(flowError === 'unknown' || flowError === 'expired') && (
        <button
          type="button"
          className={flowStyles.ctaButton}
          onClick={handleTryAgainClick}
        >
          Try again
        </button>
      )}

      <LinkBack />
    </section>
  );
}
