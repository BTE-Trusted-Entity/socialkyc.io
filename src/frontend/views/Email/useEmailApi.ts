import { useMemo } from 'react';

import { quoteEmail } from '../../../backend/email/quoteEmailApi';
import { sendEmail } from '../../../backend/email/sendEmailApi';
import { confirmEmail } from '../../../backend/email/confirmEmailApi';
import { attestEmail } from '../../../backend/email/attestEmailApi';
import { requestAttestationEmail } from '../../../backend/email/requestAttestationEmailApi';

import { bindToSession } from '../../utilities/bindToSession';

export function useEmailApi(sessionId: string) {
  return useMemo(() => {
    const sessionBound = bindToSession(sessionId);
    return {
      send: sessionBound(sendEmail),
      confirm: sessionBound(confirmEmail),
      quote: sessionBound(quoteEmail),
      requestAttestation: sessionBound(requestAttestationEmail),
      attest: sessionBound(attestEmail),
    };
  }, [sessionId]);
}
