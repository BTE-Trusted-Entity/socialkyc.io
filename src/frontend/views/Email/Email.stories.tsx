import { Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { EmailTemplate } from './EmailTemplate';
import { EmailProfile } from './Email';

export default {
  title: 'Views/Email',
  component: EmailTemplate,
} as Meta;

const actions = {
  handleSendEmail: action('sendEmail'),
  handleRequestAttestation: action('requestAttestation'),
  handleBackup: action('backup'),
  handleTryAgainClick: action('tryAgain'),
  setInputError: action('setInputError'),
};

const profileMock: EmailProfile = {
  email: 'user@example.com',
};

export function Start(): JSX.Element {
  return <EmailTemplate status="none" processing={false} {...actions} />;
}

export function EmailSent(): JSX.Element {
  return <EmailTemplate status="emailSent" processing={false} {...actions} />;
}

export function Authorizing(): JSX.Element {
  return <EmailTemplate status="authorizing" processing={false} {...actions} />;
}

export function Authorized(): JSX.Element {
  return (
    <EmailTemplate
      status="authorized"
      processing={false}
      profile={profileMock}
      {...actions}
    />
  );
}

export function QuoteOpen(): JSX.Element {
  return (
    <EmailTemplate
      status="authorized"
      processing={true}
      profile={profileMock}
      {...actions}
    />
  );
}

export function Attesting(): JSX.Element {
  return <EmailTemplate status="attesting" processing={false} {...actions} />;
}

export function Ready(): JSX.Element {
  return <EmailTemplate status="ready" processing={false} {...actions} />;
}

export function Expired(): JSX.Element {
  return (
    <EmailTemplate
      status="authorized"
      flowError="expired"
      processing={false}
      {...actions}
    />
  );
}

export function Closed(): JSX.Element {
  return (
    <EmailTemplate
      status="authorized"
      flowError="closed"
      processing={false}
      profile={profileMock}
      {...actions}
    />
  );
}

export function UnexpectedError(): JSX.Element {
  return (
    <EmailTemplate
      status="error"
      flowError="unknown"
      processing={false}
      profile={profileMock}
      {...actions}
    />
  );
}
