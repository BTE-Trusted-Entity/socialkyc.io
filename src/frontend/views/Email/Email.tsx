import { FormEvent, useCallback, useEffect, useState } from 'react';
import { IEncryptedMessage } from '@kiltprotocol/types';

import { Rejection, Session } from '../../utilities/session';
import { useValuesFromRedirectUri } from '../../utilities/useValuesFromRedirectUri';
import { InvalidEmail } from '../../../backend/email/sendEmailApi';

import { useEmailApi } from './useEmailApi';
import { AttestationStatus, EmailTemplate, FlowError } from './EmailTemplate';

export interface EmailProfile {
  Email: string;
}

interface Props {
  session: Session;
}

export function Email({ session }: Props): JSX.Element {
  const [inputError, setInputError] = useState<string>();

  const { secret } = useValuesFromRedirectUri(true);

  const initialStatus = secret ? 'authenticating' : 'none';

  const [flowError, setFlowError] = useState<FlowError>();
  const [status, setStatus] = useState<AttestationStatus>(initialStatus);
  const [processing, setProcessing] = useState(false);

  const emailApi = useEmailApi(session.sessionId);
  const [backupMessage, setBackupMessage] = useState<IEncryptedMessage>();

  const [profile, setProfile] = useState<EmailProfile>();

  useEffect(() => {
    if (!secret) {
      return;
    }
    (async () => {
      try {
        setProfile(await emailApi.confirm({ secret }));
        setStatus('authenticated');
      } catch {
        setStatus('error');
        setFlowError('expired');
      }
    })();
  }, [emailApi, secret]);

  const handleSendEmail = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();
      setProcessing(true);
      setInputError(undefined);
      setFlowError(undefined);

      const formData = new FormData(event.target as HTMLFormElement);
      const emailInput = formData.get('email') as string;

      const { wallet } = session;

      try {
        await emailApi.send({ wallet, email: emailInput.trim() });
        setStatus('emailSent');
      } catch (exception) {
        if (exception instanceof InvalidEmail) {
          setInputError('Incorrect email format, please review.');
        } else {
          console.error(exception);
          setFlowError('unknown');
          setStatus('error');
        }
      } finally {
        setProcessing(false);
      }
    },
    [emailApi, session],
  );

  const handleRequestAttestation = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();
      setProcessing(true);
      setFlowError(undefined);
      try {
        await session.listen(async (message) => {
          try {
            await emailApi.requestAttestation({ message });
            setStatus('attesting');
            setProcessing(false);

            setBackupMessage(await emailApi.attest({}));
            setStatus('ready');
          } catch (exception) {
            if (exception instanceof Rejection) {
              setFlowError('closed');
            } else {
              console.error(exception);
              setFlowError('unknown');
              setStatus('error');
            }
          }
        });

        const message = await emailApi.quote({});

        await session.send(message);
      } catch (error) {
        console.error(error);
        setStatus('error');
        setFlowError('unknown');
      } finally {
        setProcessing(false);
      }
    },
    [emailApi, session],
  );

  const handleBackup = useCallback(async () => {
    try {
      if (!backupMessage) {
        throw new Error('No backup message');
      }
      await session.send(backupMessage);
    } catch (exception) {
      if (exception instanceof Rejection) {
        return; // donâ€™t care that the user has closed the dialog
      }
      setFlowError('unknown');
      console.error(exception);
    }
  }, [backupMessage, session]);

  const handleTryAgainClick = useCallback(() => {
    setStatus('none');
    setFlowError(undefined);
  }, []);

  return (
    <EmailTemplate
      status={status}
      processing={processing}
      flowError={flowError}
      inputError={inputError}
      setInputError={setInputError}
      handleSendEmail={handleSendEmail}
      handleRequestAttestation={handleRequestAttestation}
      handleBackup={handleBackup}
      handleTryAgainClick={handleTryAgainClick}
      profile={profile}
    />
  );
}
