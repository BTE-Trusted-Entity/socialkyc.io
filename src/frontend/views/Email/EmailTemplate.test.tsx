// expect cannot be imported because of https://github.com/testing-library/jest-dom/issues/426
import { describe, it, jest } from '@jest/globals';

import { render } from '../../../testing/testing';
import '../../components/useCopyButton/useCopyButton.mock';

import { EmailTemplate } from './EmailTemplate';
import { EmailProfile } from './Email';

jest.useFakeTimers();
jest.setSystemTime(new Date('2022-01-03T12:00:00'));

const actions = {
  handleSendEmail: jest.fn(),
  handleRequestAttestation: jest.fn(),
  handleBackup: jest.fn(),
  handleTryAgainClick: jest.fn(),
  setInputError: jest.fn(),
};

const profileMock: EmailProfile = {
  email: 'user@example.com',
};

describe('EmailTemplate', () => {
  it('should match snapshot with status=none', async () => {
    const { container } = render(
      <EmailTemplate status="none" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=emailSent', async () => {
    const { container } = render(
      <EmailTemplate status="emailSent" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=authorizing', async () => {
    const { container } = render(
      <EmailTemplate status="authorizing" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=authorized', async () => {
    const { container } = render(
      <EmailTemplate
        status="authorized"
        processing={false}
        profile={profileMock}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=attesting', async () => {
    const { container } = render(
      <EmailTemplate status="attesting" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=ready', async () => {
    const { container } = render(
      <EmailTemplate status="ready" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with flowerror=expired', async () => {
    const { container } = render(
      <EmailTemplate
        status="error"
        flowError="expired"
        processing={false}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with flowError=closed', async () => {
    const { container } = render(
      <EmailTemplate
        status="none"
        flowError="closed"
        processing={false}
        profile={profileMock}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with flowError=unknown', async () => {
    const { container } = render(
      <EmailTemplate
        status="error"
        flowError="unknown"
        processing={false}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
