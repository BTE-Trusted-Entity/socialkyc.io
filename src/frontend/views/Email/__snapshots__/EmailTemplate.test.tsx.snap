// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailTemplate should match snapshot with flowError=closed 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Email Address Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <form>
      <label>
        Your email address
        <input
          class="formInput"
          maxlength="100"
          name="email"
          required=""
          type="email"
          value=""
        />
      </label>
      <output
        class="inputError"
        hidden=""
      />
      <p>
        Validity: one year (
        1/3/2023
        )
      </p>
      <button
        class="ctaButton"
        disabled=""
        type="submit"
      >
        Send email
      </button>
    </form>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`EmailTemplate should match snapshot with flowError=unknown 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Email Address Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="exclamation"
      />
      <h2
        class="heading"
      >
        Attestation error:
      </h2>
      <p
        class="message"
      >
        Something went wrong!
      </p>
      <p>
        Click “Try Again” or reload the page or restart your browser.
      </p>
    </section>
    <button
      class="ctaButton"
      type="button"
    >
      Try again
    </button>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`EmailTemplate should match snapshot with flowerror=expired 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Email Address Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="exclamation"
      />
      <h2
        class="heading"
      >
        Attestation error:
      </h2>
      <p
        class="message"
      >
        This link has expired.
      </p>
    </section>
    <button
      class="ctaButton"
      type="button"
    >
      Try again
    </button>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`EmailTemplate should match snapshot with status=attesting 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Email Address Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="spinner"
      />
      <h2
        class="heading"
      >
        Attestation process:
      </h2>
      <p
        class="message"
      >
        Anchoring credential on KILT blockchain
      </p>
      <p>
        Please leave this tab open until your credential is attested.
      </p>
    </section>
  </section>
</div>
`;

exports[`EmailTemplate should match snapshot with status=authorized 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Email Address Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <form>
      <dl
        class="profile"
      >
        <dt>
          Email:
        </dt>
        <dd>
          user@example.com
        </dd>
      </dl>
      <p>
        Validity: one year (
        1/3/2023
        )
      </p>
      <button
        class="ctaButton"
      >
        Continue in wallet
      </button>
    </form>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`EmailTemplate should match snapshot with status=authorizing 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Email Address Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="spinner"
      />
      <h2
        class="heading"
      >
        Attestation process:
      </h2>
      <p
        class="message"
      >
        Authorizing
      </p>
      <p>
        Please wait, confirming email .
      </p>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`EmailTemplate should match snapshot with status=emailSent 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Email Address Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="spinner"
      />
      <h2
        class="heading"
      >
        Attestation process:
      </h2>
      <p
        class="message"
      >
        Email sent
      </p>
      <p>
        Email sent to . Please check your inbox and spam folder and click the link.
      </p>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`EmailTemplate should match snapshot with status=none 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Email Address Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <form>
      <label>
        Your email address
        <input
          class="formInput"
          maxlength="100"
          name="email"
          required=""
          type="email"
          value=""
        />
      </label>
      <output
        class="inputError"
        hidden=""
      />
      <p>
        Validity: one year (
        1/3/2023
        )
      </p>
      <button
        class="ctaButton"
        disabled=""
        type="submit"
      >
        Send email
      </button>
    </form>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`EmailTemplate should match snapshot with status=ready 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Email Address Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section
      class="container"
    >
      <div
        class="checkmark"
      />
      <h2
        class="heading"
      >
        Attestation process:
      </h2>
      <p
        class="message"
      >
        Credential is ready
      </p>
      <p>
        We recommend that you back up your credential now.
      </p>
    </section>
    <button
      class="ctaButton"
      type="button"
    >
      Show credential in wallet
    </button>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;
